Allow parenthesis to be preserved

wenq1 opened the pull request:
Should preserve spacing now. 

```js
                let     js_source       = [
                        // `function a () {`,
                        `        testFunc (/** @type {string} */ (a), b);`,
                        `        testFunc1A (/** @type {string} */ (a), b, {c: 1, d: 1});`,
                        `        testFunc2A (/** @type {string} */ (a), b);`,
                        // `}`,
                ].join ('\n');


recastVisit (js_ast, {
                        visitCallExpression:    function (path1) {
                                if (path1.value.callee.name === 'testFunc') {
                                        return false;
                                }
                                else if (path1.value.callee.name === 'testFunc1A') {
                                        this.traverse (path1);
                                }

                                if (path1.value.callee.name === 'testFunc2A') {
                                        path1.value.arguments.unshift (astTypes.builders.identifier ('newOpt'));
                                }
                                else if (path1.value.callee.name === 'testFunc1A') {
                                        path1.value.arguments[1].name   = 'b123456789';
                                }

                                return false;
                        },

                        /**
                         * @param {*} path1
                         * @returns
                         */
                        visitObjectExpression:  function (path1) {
                                let     prop0   = path1.value.properties[0];
                                prop0.key.name  = 'c000';
                                return false;
                        },

                });
```

coderaiser commented:
Please fix lint, add test and send PR to recast so they can also comment and merge

wenq1 commented:
Fixed and test passed.

wenq1 commented:
> recast

it's based on your last PR's work on the comments section.

There's no point creating a PR until yours is merged.

coderaiser commented on test/comments.ts:
```diff
@@ -833,15 +833,15 @@ function runTestsForParser(parserId: any) {
     );
   });
-  
+
```
Would be better to add a new test in `patcher.ts`

wenq1 replied:
All 731 tests passed.

There is one "pending", which is `    - should parse and print fragments` in ` Esprima JSX Compatibility`, which has nothing to do with this PR.

coderaiser commented on lib/patcher.ts:
```diff
@@ -172,6 +172,67 @@ export function getReprinter(path: any) {
+                          }
+                  }
+              }
+              // else if (origIndex === 0) {
```
What purpose of this commented code?

wenq1 replied:
this is not a completed implementation, as not all cases are handled. However cases that are indeed handled should work as expected.

Basic idea is this: let's be greedy and find the location of the semantic element before argument `a`, in `testFunc (/** @type */ (a), b)`, in order to grep the whole `(a)` lot.

if `Identifier` a has leading comments, let's find till `*/`;
else if a is the first argument, let's find the place of the opening parenthesis (to answer your question)
else let's find the place of the previous comma Token.

Same logic goes to finding the last position of the `(a)` whole lot.

As for the failed tests on NodeJS 12 and 14, its due to `comments.at (-1)` usage. It's the same as `comments[comments.length-1]`. However, I think it should be handled by tsconfig for transpiling, not the code itself.

wenq1 commented:
Do you wish the issue be further discussed or closed? Please just reply and I'll work on it further or close the PR.

wenq1 commented:
This is my last reply for bikeshedding before I drop this commit. I run commercial project. I don't have time to waste.

> Well, right now I see 5 failed tests, better to create test in `patcher.ts`. 

Let's clear this up. I introduced new code using Array.prototype.at (). This is not supported by Node.js <= 14. I have fixed it in the last commit.

> The thing is I maintain the fork of Recast, because things go very slowly as you see. 

You don't have to be slow. They can leave a PR two weeks hanging without a single reply. Fcuk them. You don't need to behave like them.

Anyways I rebase all changes from Recast repo, because it has much more contributors, and maintainers adds some fixes and features once in a couple months. 

> I seriously doubt so.

I have a list of unmerged PR’s, and I resolve conflicts each time I rebase. So, the more your code will be written in style of other Recast code, 

> No fcuking way I write code in their fcuking stykle.

> more chances it will be merged. Maintain such a library is not fun, and even less fun, to support a fork, but most my use cases, and use cases of 🐊Putout users I handle.

You could always make it fun. In fact, the whole fcuking point of forking is to make fun. If they don't work the fun way, fcuk them, no need to bring the aweful thing into our discussion.


> Could you please tell me what is your use case while you using 🐊Putout?

Sure, we have a large code base (over a million lines of javascript, and properly documented using JSDoc which is also compatible with Typescript's typings). The back end is pure nodejs, no Deno, no Bun.js, just Node. we're happy with it without a transpilig step. We have worked around may of Microsoft's features, and doing code refactoring is non-trivial, not to mention migration into TS (which will be a horrible idea).

So, seeing things like var `a = /** type {string} */ (b);` not working is a deal breaking. And yes, this answers your previous question, how to determine leading/trailing comments. I think the /** @type {} */ is a good enough indication.

> And while your changes not in main line you should know that there is yarn aliases, and npm-patch-package, so you already can use your changes on your codebase.

Already done so. I have created a PR just with the belief that you are not a weirdo as most of the other open source project creators. You actually did something great and provided real help to others, and are not like those who masquerade their narcissism behind their "my software is optionated" flag, which to me sounds bitching.
